src/ir/principal-type.cpp

```diff
@@ -0,0 +1,1062 @@
+  if (auto* i = std::get_if<Index>(&b.exact)) {
+    if (!assignments.exacts.assign(*i, Exact)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool match(VarAssignments& assignments,
+           const VarDefHeapType& a,
+           const VarAbsHeapType& b) {
+  if (std::get_if<BottomShare>(&b.share)) {
+    return false;
+  }
+  auto* shareB = std::get_if<Shareability>(&b.share);
+  HeapType htB = b.ht.getBasic(shareB ? *shareB : a.ht.getShared());
+  if (!HeapType::isSubType(a.ht, htB)) {
+    return false;
+  }
+  if (auto* i = std::get_if<Index>(&a.exact)) {
+    if (!assignments.exacts.assign(*i, Inexact)) {
+      return false;
+    }
+  }
+  if (auto* i = std::get_if<Index>(&b.share)) {
+    if (!assignments.shares.assign(*i, a.ht.getShared())) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool match(VarAssignments& assignments,
+           const VarHeapType& a,
+           const VarHeapType& b) {
+  if (auto* i = std::get_if<Index>(&a)) {
+    return assignments.heapTypes.assign(*i, b);
+  }
+  if (auto* i = std::get_if<Index>(&b)) {
+    return assignments.heapTypes.assign(*i, a);
+  }
+  if (std::get_if<BottomHeapType>(&a)) {
+    return matchBottom(assignments, b);
+  }
+  if (std::get_if<BottomHeapType>(&b)) {
+    return false;
+  }
+  auto* absA = std::get_if<VarAbsHeapType>(&a);
+  auto* absB = std::get_if<VarAbsHeapType>(&b);
+  auto* defA = std::get_if<VarDefHeapType>(&a);
+  auto* defB = std::get_if<VarDefHeapType>(&b);
+  if (absA && absB) {
+    return match(assignments, *absA, *absB);
+  }
+  if (absA && defB) {
+    return match(assignments, *absA, *defB);
+  }
+  if (defA && absB) {
+    return match(assignments, *defA, *absB);
+  }
+  if (defA && defB) {
+    return match(assignments, *defA, *defB);
+  }
+  WASM_UNREACHABLE("unexpected variants");
+}
+
+bool match(VarAssignments& assignments, const VarRef& a, const VarRef& b) {
+  return match(assignments, a.null, b.null) && match(assignments, a.ht, b.ht);
+}
+
+// Update the assignments as necessary to make a <: b, returning true iff this
+// is possible. Otherwise the assignments may be partially modified.
+bool match(VarAssignments& assignments, const VarType& a, const VarType& b) {
+  assert(valid(a) && valid(b));
+  if (auto* i = std::get_if<Index>(&a)) {
+    return assignments.types.assign(*i, b);
+  }
+  if (auto* i = std::get_if<Index>(&b)) {
+    return assignments.types.assign(*i, a);
+  }
+
+  auto* refB = std::get_if<VarRef>(&b);
+  assert(!refB || b == canonicalizeRef(*refB));
+  auto* refA = std::get_if<VarRef>(&a);
+  assert(!refA || a == canonicalizeRef(*refA));
+
+  if (refA && refB) {
+    return match(assignments, *refA, *refB);
+  }
+
+  auto* typeA = std::get_if<Type>(&a);
+  auto* typeB = std::get_if<Type>(&b);
+
+  if (typeA && typeB) {
+    return Type::isSubType(*typeA, *typeB);
+  }
+
+  if (typeA && *typeA == Type::unreachable) {
+    assert(refB);
+    return matchBottom(assignments, *refB);
+  }
+
+  if (typeA && refB) {
```

Comment:
```suggestion
  if (typeA) {
```
At this point we've ruled out indexes, and ruled out that both are simple types, or that both are varrefs, so one must be a type and one must be a varref, hence typeA implies refB?

Reply:
Yes, but I think it's better in this case to directly spell out the cases we depend on rather than depending on implications that we know to be true. In particular, if we have to add more variants for some reason in the future, it will be easier to reason through the necessary changes with the cases explicitly spelled out.

Reply:
sgtm, but maybe comment about the redundancy, just to avoid reader confusion?